src/gui/toolbarMenubar/AbstractSliderItem.cpp

```diff
@@ -0,0 +1,194 @@
+
+    /**
+     * Listen for input events on the new slider.
+     *
+     * If this already has a slider, this disconnects listeners
+     * on events for the current slider.
+     *
+     * @param slider The new slider to listen for events on.
+     */
+    void setSlider(GtkRange* slider);
+
+    void setEnabled(bool enabled);
+
+    /**
+     * Constructs a new toolbar item and slider.
+     * Does not take ownership.
+     */
+    std::tuple<GtkToolItem*, GtkRange*> newItem(bool horizontal) const;
+
+    // Conversions to/from range_ (internal values) to external scale values:
+    double getScaleMax() const { return publicApi_->scaleFunc(range_.max); };
+    double getScaleMin() const { return publicApi_->scaleFunc(range_.min); };
+    double getFineStepSize() const { return (getScaleMax() - getScaleMin()) / range_.fineSteps; };
+    double getCoarseStepSize() const { return (getScaleMax() - getScaleMin()) / range_.coarseSteps; };
+
+public:
+    bool horizontal_ = false;
+    GtkRange* slider_ = nullptr;
+    GtkToolItem* toolItem_ = nullptr;
+    AbstractSliderItem* publicApi_;
+
+    const SliderRange range_;
+};
+
+AbstractSliderItem::~AbstractSliderItem() = default;
+AbstractSliderItem::AbstractSliderItem(std::string id, ActionHandler* handler, ActionType type, SliderRange range):
+        AbstractToolItem{std::move(id), handler, type}, pImpl{std::make_unique<Impl>(this, range)} {}
+
+void AbstractSliderItem::configure(GtkRange* slider, bool isHorizontal) const {
+    double fineStepSize = pImpl->getFineStepSize();
+    double coarseStepSize = pImpl->getCoarseStepSize();
+
+    gtk_range_set_increments(slider, fineStepSize, coarseStepSize);
+
+    if (isHorizontal) {
+        gtk_widget_set_size_request(GTK_WIDGET(slider), 120, 16);
+    } else {
+        gtk_widget_set_size_request(GTK_WIDGET(slider), 16, 120);
+    }
+
+    gtk_scale_set_draw_value(GTK_SCALE(slider), true);
+    gtk_widget_set_can_focus(GTK_WIDGET(slider), false);
+}
+
+auto AbstractSliderItem::createItem(bool horizontal) -> GtkToolItem* {
+    pImpl->horizontal_ = horizontal;
+    auto [toolItem, slider] = pImpl->newItem(horizontal);
+
+    pImpl->toolItem_ = toolItem;
+    g_object_ref(toolItem);
+    pImpl->setSlider(slider);
+
+    if (GTK_IS_TOOL_ITEM(toolItem)) {
+        gtk_tool_item_set_homogeneous(GTK_TOOL_ITEM(toolItem), false);
+    }
+    if (GTK_IS_TOOL_BUTTON(toolItem) || GTK_IS_TOGGLE_TOOL_BUTTON(toolItem)) {
+        g_signal_connect(toolItem, "clicked", G_CALLBACK(&toolButtonCallback), this);
+    }
+
+    return toolItem;
+}
+
+auto AbstractSliderItem::createTmpItem(bool horizontal) -> GtkToolItem* {
+    auto [item, slider] = pImpl->newItem(horizontal);
+    g_object_ref(item);
+
+    gtk_widget_show_all(GTK_WIDGET(item));
+    return item;
+}
+
+// TODO(personalizedrefrigerator): Remove this (currently required by AbstractToolItem
+auto AbstractSliderItem::newItem() -> GtkToolItem* { return createTmpItem(true); }
+
```

Comment:
Currently, `AbstractToolItem` requires an implementation of `newItem` in addition to `createItem` and `createTmpItem`. 

In `newItem` it is not specified whether the item is `horizontal` or not and `newItem` only returns one thing, a `GtkToolItem` (no pointers to internal widgets, e.g. a slider). As such, it isn't clear whether `newItem` is intended to modify the state of `this`, or whether it should create an empty toolbar item...